### PR TITLE
Feature/api work package update

### PR DESCRIPTION
**Pull request by mtakac2**

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage. <a href='http://feature-api-work-package-update.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

File: lib/api/v3/work_packages/work_package_model.rb [hunk 1]
```diff
@@ -157,8 +158,21 @@ def is_closed
           work_package.closed?
         end
 
+        def custom_properties=(values)
+          binding.pry
```
Comment (myabc) [inline lib/api/v3/work_packages/work_package_model.rb]: @mtakac debug code shouldn't be here.

File: lib/api/v3/work_packages/work_packages_api.rb [hunk 2]
```diff
@@ -49,6 +49,17 @@ class WorkPackagesAPI < Grape::API
               @representer.to_json
             end
 
+            patch do
+              authorize(:edit_work_packages, context: @work_package.project)
+              @representer.from_json(params.except(:route_info, :id).to_json)
```
Comment (myabc) [inline lib/api/v3/work_packages/work_packages_api.rb]: @mtakac why can't we just pass the request body?

File: lib/api/v3/work_packages/work_package_model.rb (hunk 1 above)
Comment (mtakac2) [inline lib/api/v3/work_packages/work_package_model.rb]: @myabc uuugrh I'll add some git hooks to check for bindings if you don't mind.

File: lib/api/v3/work_packages/work_packages_api.rb (hunk 2 above)
Comment (mtakac2) [inline lib/api/v3/work_packages/work_packages_api.rb]: @myabc sure, I'll change it to request.POST

File: lib/api/v3/work_packages/work_package_model.rb [hunk 3]
```diff
@@ -79,7 +80,7 @@ def type
         end
 
         def type=(value)
-          type = Type.find(:first, conditions: ['name ilike ?', value])
+          type = Type.find(:first, conditions: ['UPPER(name) = UPPER(?)', value])
```
Comment (myabc) [inline lib/api/v3/work_packages/work_package_model.rb:83]: @mtakac why do we need case insensitivity here?

Until we move to Rails 4, why not just `Type.find_by_name`?

File: lib/api/decorators/resource.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
+#-- encoding: UTF-8
```
Comment (myabc) [inline lib/api/decorators/resource.rb]: Where's this being used?

File: lib/api/v3/work_packages/work_package_model.rb (hunk 3 above)
Comment (mtakac2) [inline lib/api/v3/work_packages/work_package_model.rb:83]: @myabc I don't understand the question. Is Type.find_by_name case insensitive?

Comment (myabc) [inline lib/api/v3/work_packages/work_package_model.rb:83]: @mtakac No, it's not. My original question:

> why do we need case insensitivity here?

Comment (mtakac2) [inline lib/api/v3/work_packages/work_package_model.rb:83]: Well, we don't but it's probably much simpler for the user if he doesn't have to match the case in the request.

Comment (myabc) [inline lib/api/v3/work_packages/work_package_model.rb:83]: @mtakac that's inconsistent with the behaviour elsewhere in the application.

Comment (mtakac2) [inline lib/api/v3/work_packages/work_package_model.rb:83]: It's consistent across the API. I am not aware of any similar situation elsewhere (correct me if I am wrong). My point is I don't think we should force API user to know if the name of a type is 'Bug' or 'bug'. I think it should work with both. @ulferts @manwithtwowatches @hschink any preferences on this one?

Comment (myabc) [inline lib/api/v3/work_packages/work_package_model.rb:83]: > My point is I don't think we should force API user to know if the name of a type is 'Bug' or 'bug'. 

`Status` is handled case-sensitively elsewhere in the system. For example, it's perfectly possible to do this:

``` ruby
Status.create(name: 'Bug')
# => #<Status id: 1, name: "Bug",…
Status.create(name: 'bug')
# => #<Status id: 2, name: "bug",…
```

If we're not forcing them, how would we determine which one they want?

Regardless of this, SQL conditions do not belong in the `WorkPackageModel`. You should create a finder method in the relevant ActiveRecord model class.

**Pull request merged**